File: Python/Smith_Bryan_BankReport.py
#!/usr/bin/python
import sys

##########################################################
# CSC 4200-01 - Programming Languages
# Name: Bryan Smith
# Date: 2/21/14
# Description: Print bank statment based on data read from
#              file
# Paramaters:
#   fileLoc - Path to the file to read.
##########################################################

#Class that represents a users bank account
class Account:
    ID=-1
    initialBalance=0
    balance=0
    transactions=[]
    numbOfOverdrafts=0
    overdraftedLast=False
    overdraftCharge = 10.0

    def __init__(self, ID, balance, odc=10.0):
        self.ID = ID
        self.balance = balance
        self.initialBalance = balance
        self.numbOfOverdrafts = 0
        self.overdraftCharge = odc
        self.overdraftedLast = False

    #Calculates the final balance and any overdrafts
    def calculateBalance(self):
        self.balance=self.initialBalance
        for t in self.transactions:
            if (isinstance(t,Deposit)):
                self.balance+=t.getAmount()
            else:
                self.balance-=t.getAmount()
            if (self.balance < 0 and self.overdraftedLast==False):
                self.numbOfOverdrafts+=1
                self.balance-=self.overdraftCharge
                self.overdraftedLast=True
            elif(self.balance >= 0 and self.overdraftedLast==True):
                self.overdraftedLast=False

    #Adds a transaction to the account
    def addTransaction(self,t):
        self.transactions.append(t)

    #Removes a transaction from the account  
    def removeTransaction(self,t):
        self.transactions.remove(t)

    #Parses data from a string into data to load into the account
    def parseToAccount(self,lineList):
        try:
            self.ID=lineList[0].strip().split(":")[1].strip()
            self.initialBalance=float(lineList[1].strip().split(":")[1])
            self.balance=self.initialBalance
            for i in range(2,len(lineList)):
                tmpList=lineList[i].strip().split(":")
                if (tmpList[0].lower().strip().startswith("w")):
                    self.addTransaction(Withdrawal(tmpList[1].strip(),float(tmpList[2].strip())))
                else:
                    self.addTransaction(Deposit(tmpList[1].strip(),float(tmpList[2].strip())))
        except:
            print("Error: Could not parse account.")
            sys.exit(2)

    #Print withdrawal summary
    def printWithdrawals(self):
        totalWithdrawalAmount=0.0
        tmpBalance=self.initialBalance
        isFirstListed=True
        print("Summary of Withdrawals:	 amount	        Running Balance")
        print("-------------------------------------------------------")
        for t in self.transactions:
            if (isinstance(t,Deposit)):
                print("--deposit(see below)")
                tmpBalance+=t.getAmount()
            else:
                t.printTransaction(False,-isFirstListed)
                isFirstListed=False
                totalWithdrawalAmount+=t.getAmount()
                if(tmpBalance <0):
                    self.overdraftedLast=True
                else:
                    self.overdraftedLast=False
                tmpBalance-=t.getAmount()
                if(tmpBalance >0):
                   print("{0:>24}".format(tmpBalance))
                elif (tmpBalance <0 and self.overdraftedLast == False):
                    sys.stdout.write("{0:>24}".format("( {:.2f})".format(-tmpBalance)))
                    sys.stdout.write("*\n")
                    
        print("-------------------------------------------------------")
        print("{0:<21} {1:>9}".format("Total Withdrawals","${:.2f}".format(totalWithdrawalAmount)))

    #Print Deposits summary
    def printDeposits(self):
        totalDepositAmount=0.0
        isFirstListed=True
        print("\nSummary of Deposits:")
        print("-------------------------------")
        for t in self.transactions:
            if (isinstance(t,Deposit)):
                t.printTransaction(True,-isFirstListed)
                isFirstListed=False
                totalDepositAmount+=t.getAmount()
                    
        print("\n-------------------------------")
        print("{0:<16} {1:>14}".format("Total Deposits","${:.2f}".format(totalDepositAmount)))

    #Print the whole bank statement
    def __str__( self ):
        print("Statement for Account: %s\n" %(self.ID))
        print("Beginning Balance: $%.2d \n" %(self.initialBalance))
        self.printWithdrawals()
        self.printDeposits()

        if (self.numbOfOverdrafts>0):
            print("\n*Overdraft Fees")
            print("--------------")
            for i in range(self.numbOfOverdrafts):
                print("$%.2d"%(self.overdraftCharge))
            print("--------------")
        if (self.balance <0):
            print("{0:<16} {1:>14}".format("\nEnding Balance:","( ${:.2f})".format(-self.balance)))
        else:
            print("{0:<16} {1:>14}".format("\nEnding Balance:","${:.2f}".format(self.balance)))
        return ""
    
#Class that represents a bank transaction
class Transaction:
    source = "Default Source"
    amount = 0.0

    def __init__(self, s, a):
        self.source = s;
        self.amount = a;

    def printTransaction(self,newline, cur):
        if (newline and cur):
            sys.stdout.write("{0:<21} {1:>9}".format(self.source," ${:.2f}\n".format(self.amount)))
        elif(cur):
            sys.stdout.write("{0:<21} {1:>9}".format(self.source,"${:.2f}".format(self.amount)))
        else:
            sys.stdout.write("{0:<21} {1:>9}".format(self.source,"{:.2f}".format(self.amount)))

    def getAmount(self):
        return self.amount

#Child class that represents a withdrawal transaction
class Withdrawal(Transaction):
        pass

#Child class that represents a deposit transaction
class Deposit(Transaction):
        pass


#Reads in a file returning the string of what was read.
def readFile(fileLoc):
    try:
        content = []
        with open(fileLoc) as f:
            content = f.readlines()
        return content
    except:
        print("Error Reading file. Does it exist?")
        sys.exit(2)

#If there is a command line argument, use it as the input file path
if (len(sys.argv)>1):
    fileLoc=str(sys.argv[1])
else:
    fileLoc="TestFile.txt"

#Create the bank account, parse the file to load in the data, calculate and print.
user = Account(-1,0.0)
user.parseToAccount(readFile(fileLoc))
user.calculateBalance()
print user

raw_input("\n\nPress the enter key to exit.") #Wait on user so data is displayed

File: Perl/Smith_Bryan_Bankreport.pl
#!/usr/bin/perl
##########################################################
# CSC 4200-01 - Programming Languages
# Name: Bryan Smith
# Date: 2/28/14
# Description: Print bank statement based on data read from
#              file
# Parameters:
#   fileLoc - Path to the file to read.
##########################################################

use constant { true => 1, false => 0 };

#Check for command line file location. If not specified use default.
if (scalar @ARGV >= 1) {
 $fileLoc = @ARGV[0]; 
} else {
 $fileLoc = "TestData.txt"; 
}

#Open file based on supplied file location and store the variable.
open(DATA, "<".$fileLoc) or die "Error reading file ".$fileLoc." Does it exist?";
@fileContents = <DATA>;
close DATA or warn "Error closing input file.";

#Main setup
#Parses the file, loading the content into an Account object. Then prints out the statement that object generates.
eval {
  @line1 = split(/:/,@fileContents[0]);
  @line2 = split(/:/,@fileContents[1]);
  $curAccount = new Account(trim(@line1[1]),trim(@line2[1]),10.0);
  
  for($i=2;$i<scalar @fileContents;$i=$i+1) {
    @split = split(/:/,@fileContents[$i]);
    @split[0] = substr(@split[0], 0, 1);
    @split[0] =~ tr/[A-Z]/[a-z]/;
    if (@split[0] eq 'd') {
        $curAccount->addTransaction(new Deposit(trim(@split[1]),trim(@split[2])));
    }elsif (@split[0] eq 'w') {
        $curAccount->addTransaction(new Withdrawal(trim(@split[1]),trim(@split[2])));
    }
  }
  $curAccount->calculateBalance();
  $curAccount->print();
  1;
} or do {
    my $e = $@;
    print("Error: $e\n");
};

#Trim leading and trailing space
sub trim {
    (my $s = $_[0]) =~ s/^\s+|\s+$//g;
    return $s;        
}

########################
# Account Class
########################
package Account;
sub new
{

    my $self = {
        _class => shift,
        _id => shift,
        _initialBalance => shift,
        _balance => _initialBalance,
        _numberOfOverdrafts => 0,
        _overdraftedLast => false,
        _overdraftCharge => shift,
        _transactions => [],
    };

    bless $self;
    return $self;
}

#Adds a transaction to the account
sub addTransaction {
  my ( $self, $t ) = @_;
  $trans = $t if defined($t);
  push (@{$self->{'_transactions'}}, $trans);    
}

#Calculates the final balance and any overdrafts
sub calculateBalance {
    my ( $self, $t ) = @_;
    $self->{_balance}=$self->{_initialBalance};
    foreach $t (@{$self->{'_transactions'}}) { 
          if ($t->{'_class'} eq "Deposit") {
              $self->{_balance}+=$t->getAmount();
          } else {
              $self->{_balance}-=$t->getAmount();
          }
          if ($self->{_balance} < 0 and $self->{_overdraftedLast} eq false) {
              $self->{_numberOfOverdrafts}++;
              $self->{_overdraftedLast}=true;
          } elsif ($self->{_balance} >= 0 and $self->{_overdraftedLast} eq true) {
             $self->{_overdraftedLast}=false;
          }
    }
}

#Prints the entire account statement
sub print {
     my( $self ) = @_;
     
     printf("Statement for Account: %s\n",$self->{'_id'});
     printf("\nBeginning Balance: \$%.2f \n\n",$self->{'_initialBalance'});
     $self->printWithdrawals();
     $self->printDeposits();
     
     if ($self->{_numberOfOverdrafts}>0) {
            print("\n*Overdraft Fees\n");
            print("--------------\n");
            for ($i=0;$i<$self->{'_numberOfOverdrafts'};$i++){
                printf("\$%.2d\n",$self->{'_overdraftCharge'});
                $self->{_balance}=($self->{_balance}-$self->{_overdraftCharge});
            }
            print("--------------\n");
      }
      if ($self->{_balance} >= 0) {
        printf("%-21s %12s","\nEnding Balance:",sprintf("\$%.2f\n\n",$self->{_balance}));
      } else {
         printf("%-21s %12s","\nEnding Balance:",sprintf("( \$%.2f)\n\n",-$self->{_balance})); 
      }
     

}

#Prints the withdrawals summary
sub printWithdrawals {
     my( $self) = @_;
     $totalWithdrawalAmount=0.0;
     $tmpBalance=$self->{_initialBalance};
     $isFirstListed=true;
     
     print("Summary of Withdrawals:  amount         Running Balance\n");
     print("-------------------------------------------------------\n");
     
     foreach $t (@{$self->{'_transactions'}}) {
         if ($t->{'_class'} eq "Deposit") {
            print("--deposit(see below)");
            $tmpBalance+=$t->getAmount();
         } else {
             $t->print($isFirstListed,false);
             $isFirstListed=false;
             $totalWithdrawalAmount+=$t->getAmount();
             if($tmpBalance <0) {
                    $self->{_overdraftedLast}=true;
             } else {
                    $self->{_overdraftedLast}=false;
             }
             $tmpBalance-=$t->getAmount();
              if($tmpBalance > 0) {
                  printf("%24s",sprintf("%.2f",$tmpBalance));
              } elsif ($tmpBalance <0 and $self->{_overdraftedLast} eq false){
                  printf("%25s",sprintf("( %.2f)",-$tmpBalance));
                  print("*");
              }
         }
         print("\n");
     }
     print("-------------------------------------------------------\n");
     printf("%-21s %9s","Total Withdrawals",sprintf("\$%.2f",$totalWithdrawalAmount));
     print("\n");
}

#Prints the deposits summary
sub printDeposits {
     my( $self) = @_;
     $totalDepositAmount=0.0;
     $isFirstListed=true;
     
     print("\nSummary of Deposits:\n");
     print("-------------------------------\n");
     
     foreach $t (@{$self->{'_transactions'}}) {
         if ($t->{'_class'} eq "Deposit") {
            $t->print($isFirstListed,false);
            $isFirstListed=false;
            $totalDepositAmount+=$t->getAmount();
            print("\n");
         }
         
     }
     print("-------------------------------\n");
     printf("%-21s %9s","Total Deposits",sprintf("\$%.2f",$totalDepositAmount));
     print("\n");
}
1;


########################
# Transaction Class
########################
package Transaction;
sub new
{

    my $self = {
        _class => shift,
        _source => shift,
        _amount => shift,
    };
    bless $self;
    return $self;
}

sub setSource {
    my ( $self, $source ) = @_;
    $self->{_source} = $source if defined($source);
    return $self->{_source};
}

sub getSource {
    my( $self ) = @_;
    return $self->{_source};
}

sub setAmount {
    my ( $self, $amount ) = @_;
    $self->{_amount} = $amount if defined($amount);
    return $self->{_amount};
}

sub getAmount {
    my( $self ) = @_;
    return $self->{_amount};
}

sub print {
    my( $self,$cur,$newline ) = @_;
    $source = $self->getSource();
    $amount = $self->getAmount();
    if ($cur eq true) {
        printf("%-21s %9s",$source,sprintf("\$%.2f",$amount));
    } else {
        printf("%-21s %9s",$source,sprintf("%.2f",$amount));
    }
    if ($newline eq true) {print "\n";}
}

1;


########################
# Withdrawal Class
########################
package Withdrawal;
use base 'Transaction';
use strict;
our @ISA = qw(Transaction);
1;   


########################
# Deposit Class
########################
package Deposit;
use base 'Transaction';
use strict;
our @ISA = qw(Transaction);
1;   
